elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore.rb

```diff
@@ -43,97 +46,156 @@ module HTTP
+              sniffer_class = options[:sniffer_class] ? options[:sniffer_class] : ::Elasticsearch::Transport::Transport::HTTP::Manticore::ManticoreSniffer
+              raise ArgumentError, "Sniffer class #{sniffer_class} must be a ManticoreSniffer!" if sniffer_class.nil? || !sniffer_class.ancestors.include?(::Elasticsearch::Transport::Transport::HTTP::Manticore::ManticoreSniffer)
+              @sniffer = sniffer_class.new(self, logger)
+              @sniffer.sniff_every(options[:sniffer_delay] || 5) do |urls|
+                logger.info("Will update internal host pool with #{urls.inspect}")
+                @pool.update_urls(urls)
               end
-              Response.new resp.code, resp.read_body, resp.headers
             end
           end
 
-          # Builds and returns a collection of connections.
-          # Each connection is a Manticore::Client
-          #
-          # @return [Connections::Collection]
-          #
-          def __build_connections
-            @request_options = {}
+          # Sniff (if enabled) to get the newest list of hosts
+          # then attempt to resurrect any dead URLs
+          def reload_connections!
+            if options[:sniffing]
+              @pool.update_urls(@sniffer.hosts)
+            end
+            @pool.resurrect_dead!
+          end
+
+          def perform_request(method, path, params={}, body=nil)
+            with_request_retries do
+              body = __convert_to_json(body) if body
+              enriching_response(method, path, params, body) do
+                url, response = @pool.perform_request(method, path, params, body)
 
-            if options.key?(:headers)
-              @request_options[:headers] = options[:headers]
+                # Raise an exception so we can catch it for `retry_on_status`
+                __raise_transport_error(response) if response.status.to_i >= 300 && @retry_on_status.include?(response.status.to_i)
+                [url, response]
+              end
             end
+          end
 
-            Connections::Collection.new \
-              :connections => hosts.map { |host|
-                host[:protocol]   = host[:scheme] || DEFAULT_PROTOCOL
-                host[:port]     ||= DEFAULT_PORT
+          # This takes a host string to aid in debug logging
+          def with_request_retries
+            tries = 0
+            begin
+              tries += 1
+              yield
+            rescue ::Elasticsearch::Transport::Transport::ServerError => e
+              if @retry_on_status.include?(e.response.status)
+                logger.warn "[#{e.class}] Attempt #{tries} to get response from #{url}" if logger
+                logger.debug "[#{e.class}] Attempt #{tries} to get response from #{url}" if logger
+                if tries <= max_retries
+                  retry
+                else
+                  logger.fatal "[#{e.class}] Cannot get response from #{url} after #{tries} tries" if logger
```

Comment:
I'd call this error instead? fatal are usually reserved for errors that can lead the whole application to terminate